### PR TITLE
Split ext theory to own file and document

### PR DESCRIPTION
**Pull request by ajreynol**

### CODE REVIEW

File: src/theory/ext_theory.h [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+                          Node term,
+                          std::vector<Node>& exp,
+                          bool useCache = false);
+  /** clear the cache for getSubstitutedTerm */
+  void clearCache();
+  /** doInferences
+   *
+   * This function performs "context-dependent simplification". The method takes
+   * as input:
+   *  effort: an identifier used to determine which terms with reduce and the
```
Comment (mpreiner) [inline src/theory/ext_theory.h]: ... which terms _we_ reduce ... ?

File: src/theory/ext_theory.h [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+  /**
+   * Calls the above function, where terms is getActive(), the set of currently
+   * active terms.
+   */
+  bool doInferences(int effort, std::vector<Node>& nred, bool batch = true);
+  /** doReductions
+   *
+   * This method has the same interface as doInferences. In contrast to
+   * doInfereces, this method will send reduction lemmas of the form ( t = t' )
+   * where t is in terms, t' is equivalent, reduced term.
```
Comment (mpreiner) [inline src/theory/ext_theory.h]: ... in terms _and_ t' is _an_ equivalent ... ?

File: src/theory/ext_theory.h [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  bool sendLemma(Node lem, bool preprocess = false);
+  /** reference to the underlying theory */
+  Theory* d_parent;
+  /** the true node */
+  Node d_true;
+  /** extended function terms, map to whether they are active */
+  NodeBoolMap d_ext_func_terms;
+  /**
+   * The set of terms from d_ext_func_terms that are SAT-context-independent
+   * inactive. For instance, a term t if SAT-context-independent inactive if
```
Comment (mpreiner) [inline src/theory/ext_theory.h]: ... a term t _is_ SAT-context...

File: src/theory/ext_theory.cpp [hunk 4]
```diff
@@ -0,0 +1,542 @@
+/*********************                                                        */
+/*! \file theory.cpp
```
Comment (mpreiner) [inline src/theory/ext_theory.cpp]: ext_theory.cpp

File: src/theory/ext_theory.cpp [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Base for theory interface.
+ **
+ ** Base for theory interface.
+ **/
```
Comment (mpreiner) [inline src/theory/ext_theory.cpp]: Add a better description here.

File: src/theory/ext_theory.h [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Base of the extended theory interface.
+ **
+ ** Base of the extended theory interface.
```
Comment (mpreiner) [inline src/theory/ext_theory.h]: Can you summarize in one or two sentences what this class is about? You can copy paste it to the cpp header.

**Comment by ajreynol** [comment]
@mpreiner , anything left to do on this one?
